### ISSUE TITLE
Flink: get duplicate rows when sync CDC data by FlinkSQL

### ISSUE THREAD

**Issue by Reo-LEI** [apache/iceberg#2918]
### Duplicate Rows Problem

Recently, I trying use FlinkSQL to synchronize MySQL to iceberg. Saving MySQL/PostgreSQL to iceberg is very common case in the construction of data warehouses.  But I got duplicat rows which have same primary key in iceberg table but not in mysql table. I will illustrate the reason of this problem as follow.

Assume we have a sample MySQL table which just has one record:

id | data
---|---
1| aaa

And we define an ETL job by FlinkSQL as follow:
```
CREATE TABLE `mysql_table` (
    `id` INT, 
    `data` STRING
)
WITH (
    'connector' = 'mysql-cdc', 
    'database-name' = 'xxx', 
    ...
)

CREATE TABLE `iceberg_table` (
    `id` INT, 
    `data` STRING
    PRIMARY KEY(id) NOT ENFORCED
)
WITH (
    'connector' = 'iceberg', 
    'catalog-type' = 'hive', 
    'catalog-name' = 'iceberg', 
    'catalog-database' = 'xxx', 
    'format.version' = '2'
    ...
)

INSERT INTO `iceberg_table``
SELECT `id`, `data` FROM `mysql_table`
```

If we don't setting the default parallelism of job, we will get the job graph like this:
```
FlinkCDCSource ---forward---> Filter ---forward---> IcebergStreamWriter ---forward---> IcebergFilesCommitter
```
That will be work fine and not duplicat rows. because all opr parallelism is 1, and all cdc data will be distributed to one IcebergStreamWriter.

But once we setting the default parallelism of job by `SET table.exec.resource.default-parallelism=3`, we will get another job graph as follow:
```
                              +---> Filter ---forward---> IcebergStreamWriter-1 ---+
FlinkCDCSource ---rebalance---+---> Filter ---forward---> IcebergStreamWriter-2 ---+---rebalance---> IcebergFilesCommitter
                              +---> Filter ---forward---> IcebergStreamWriter-3 ---+
```
Notice the `Filter` opr parallelism is follow the job default parallelism and increase to 3.  For iceberg `FlinkSink`, due to the `writeParallelism` is not be set so the `IcebergStreamWriter` parallelism will follow the `rowDataInput`(`Filter`) parallelism as 3.
The `FlinkCDCSource` parallelism will be alway 1, because binlog data need to be send by serial.  Consider Flink use `rebalance` as default shuffle strategy. Now we can see the CDC data will be rebalance to three different `Filter` and then emit different `IcebergStreamWriter`. 

At this time, If we insert one row and update it as follow, we will get an duplicate row.
```
INSERT INTO `mysql_table`(id, data) VALUES(2, 'bbb');
UPDATE `mysql_table` SET data = 'xxx' WHERE id = '2';
```
MySQL:

id | data
---|---
1| aaa
2| xxx

Iceberg:

id | data
---|---
1| aaa
2| bbb
2| xxx

We got the duplicate row of `<2, 'bbb'>` because the change log data wil be rebalance to different `IcebergStreamWriter`. 
For the first `INSERT` SQL, we will got a +I record, and that will be send to the `IcebergStreamWriter-1`.
For the second `UPDATE` SQL, we will got a -U record and a +U record. The -U record will be send to the `IcebergStreamWriter-2` and +U record wil be rebalance to the `IcebergStreamWriter-3` 
Due to `IcebergStreamWriter-2` have not any data `INSERT` before, it's insertedRowMap is empty, and then the pos-delete(-U) will be ignore and will not write to delete file. 
Finally we got an duplicate row `<2, 'bbb'>`.

### Disorder Problem
The reason for duplicat rows is as mentioned above, but when I try to fix that by add keyBy before `IcebergStreamWriter`, I encounter another problem.
For FlinkSQL, flink will append a `Filter` opr before call `IcebergTableSink.getSinkRuntimeProvider` to prevenet send null data to downstream.  So, when we get the `rowDataInput` from flink in `FlinkSink`, `rowDataInput` definitely contain a `Filter` opr which parallelism is follow the job default parallelism as above. 

In our case, if we just keyBy `equalityFields` before add `IcebergStreamWriter`, we will get the job graph as follow: 
```
                              +---> Filter ---+      +---> IcebergStreamWriter-1 ---+
FlinkCDCSource ---rebalance---+---> Filter ---+-hash-+---> IcebergStreamWriter-2 ---+---rebalance---> IcebergFilesCommitter
                              +---> Filter ---+      +---> IcebergStreamWriter-3 ---+
```
Notice the CDC data come from `FlinkCDCSource` will be rebalance to `Filter` immediately.  Even if all CDC data which have same key will be emit to same `IcebergStreamWriter`, but that will make disorder between CDC data(e.g. same IcebergStreamWriter will receive +U before -U). And finally we will get a lot of strange result. 

### Goal
In summary, I think iceberg should ensure that data with the same primary key will be updated and not duplicat rows when we sink CDC/UPSERT data to iceberg. 
I trying to fix this problem by chaining the `Source` and `Filter` opr and add keyBy before `IcebergStreamWriter` when use FlinkSQL. Finally I want to got a correct job graph like this:
```
                                             +---> IcebergStreamWriter-1 ---+
FlinkCDCSource ---forward---> Filter ---hash-+---> IcebergStreamWriter-2 ---+---rebalance---> IcebergFilesCommitter
                                             +---> IcebergStreamWriter-3 ---+
```

**Comment by kbendick** [comment]
Can you show an example diagram where the filter / FlinkCDCSource has a parallelism > 1? Right now, all of your examples have parallelism of 1 for the `Source` and the `Filter`.

Also, correct me if I'm wrong, but the `Filter` operation isn't _required_ to be part of the job's DAG at all, right?

**Comment by stevenzwu** [comment]
@Reo-LEI great job on the issue description. thank you for the effort.

> The FlinkCDCSource parallelism will be alway 1, because binlog data need to be send by serial. Consider Flink use rebalance as default shuffle strategy. Now we can see the CDC data will be rebalance to three different Filter and then emit different IcebergStreamWriter.

I am not familiar with SQL or CDC parts of Flink. So the binlog is not first shipped to Kafka first here, right? Flink CDC source directly read binlog from MySQL?

Personally, I don't see too much benefit of just scale up the parallelism for the IcebergStreamWriter operator. All the extra network shuffle/rebalance can be expensive too. I do see the need to scale up the job parallelism with higher traffic. hence I am wondering if we can also scale up the Flink CDC source operator along with other operators so that we can keep the chaining behavior.

As you said already, this only works if the hash shuffle (based on the primary key) is the only network shuffle before the IcebergStreamWriter operator. That is why I have some reservations of putting this in the Flink Iceberg sink. This won't work if there is some other network shuffle before the Flink sink, which is outside of the control of the sink. Hence I am wondering if this should be handled outside of the Flink sink and we should just document the expected behavior of the input data to the Flink Iceberg sink.

**Comment by Reo-LEI** [comment]
@stevenzwu @kbendick Thanks for feed back let me know yours concern. I think I could answer your questions in here.

Although I encounter this problem when use FlinkSQL to sync cdc data to iceberg, but I think we should keep consider more general situaction about we write cdc/upsert stream data to iceberg table, such as user custom define source which produce cdc/upsert stream data or flink table retract/upsert data, and not only in flink sql but also in flink streaming. So I'd like to discuss all situaction of that.

### Discussion of current iceberg
suppose we got an `ChangelogStream` need to write to iceberg table. Consider `IcebergStreamWriter` parllelism could be config by `writeParallelism`, so we got different parllelism variate as follow:
```
x: changelog-stream-parallelism
y: job-default-parallelism
z: iceberg-writer-parallelism
```
#### For FlinkSQL
**Case1:** x == y == z          
```       
ChangelogStream-x ---forward---> Filter-y ---forward---> IcebergStreamWriter-z ---forward---> IcebergFilesCommitter
```

**Case2**: x != y == z    
```            
ChangelogStream-1 ---+           +---> Filter-1 ---forward---> IcebergStreamWriter-1 ---+
       ...           +-rebalance-+        ...                                           +-rebalance---> IcebergFilesCommitter
ChangelogStream-x ---+           +---> Filter-y ---forward---> IcebergStreamWriter-z ---+
```

**Case3**: x == y != z               
``` 
                                                       +---> IcebergStreamWriter-1 ---+
ChangelogStream-x ---forward---> Filter-y ---rebalance-+                              +-rebalance---> IcebergFilesCommitter
                                                       +---> IcebergStreamWriter-z ---+
```

**Case4**: x != y != z                
```
ChangelogStream-1 ---+           +---> Filter-1 ---+           +---> IcebergStreamWriter-1 ---+
       ...           +-rebalance-+        ...      +-rebalance-+            ...               +-rebalance---> IcebergFilesCommitter
ChangelogStream-x ---+           +---> Filter-y ---+           +---> IcebergStreamWriter-z ---+
```

In the above case, except for case 1, we can found there are at lease once `rebalance` between `ChangelogStream` and `IcebergStreamWriter`, that will make the change log records (which have same primary key) distributed to different `IcebergStreamWriter` or disorder as i said before. Finally, we will get duplicate rows or other strange result(especially Case4).

#### For FlinkStreaming
The discussions for FlinkStreaming are same as FlinkSQL, just replace `Filter` as `Map`(the `Map` opr is come from Iceberg `FlinkSink`, that is use to convert inpurt record to `RowData` type).

### Solution
In general，we could not decide the parallelism of `ChangelogStream`, and we don’t know user will how to adjust the job and writer parallelism to deal with higher traffic. In my opinion, Iceberg should ensure result correctness in all case, no matter how the user configures the parallelism of job or opr.

So in #2898, I add `keyBy` between `Filter/Map` and `IcebergStreamWriter`, that can ensure all same primary key record will be distributed to same `IcebergStreamWriter` to ensure result correctness. And I trying to set `Filter/Map` parallelism follow its input and chainig its input opr to prevent disorder problem. 

Finally, in all parallelism case as above, we can definitively get correct DAG and ensure result correctness.
```
ChangelogStream-1 ---forward---> Filter-1 ---+      +---> IcebergStreamWriter-1   ---+
      ...                          ...       +-hash-+             ...              +-rebalance---> IcebergFilesCommitter
ChangelogStream-x ---forward---> Filter-x ---+      +---> IcebergStreamWriter-y/z ---+
```

**Comment by Reo-LEI** [comment]
> Also, correct me if I'm wrong, but the `Filter` operation isn't required to be part of the job's DAG at all, right?
I think `Filter` is still required, because we need to filter the null record before FlinkSink.

**Comment by Reo-LEI** [comment]
> I am not familiar with SQL or CDC parts of Flink. So the binlog is not first shipped to Kafka first here, right? Flink CDC source directly read binlog from MySQL?

Yes, you are right, flink-cdc-connector will directly read binlog from MySQL.

> Personally, I don't see too much benefit of just scale up the parallelism for the IcebergStreamWriter operator. All the extra network shuffle/rebalance can be expensive too. I do see the need to scale up the job parallelism with higher traffic. hence I am wondering if we can also scale up the Flink CDC source operator along with other operators so that we can keep the chaining behavior.

Actually I encountered this problem when I scale up the job parallelism, and that is the Case2 above. In some situation, we can not scale up the cdc source opr, such as flink-cdc-connector the source parallelism will always be 1. And I don't think we should not depend on keep src parallelism equal to FlinkSink to guarantee result correctness.

> As you said already, this only works if the hash shuffle (based on the primary key) is the only network shuffle before the IcebergStreamWriter operator. That is why I have some reservations of putting this in the Flink Iceberg sink. This won't work if there is some other network shuffle before the Flink sink, which is outside of the control of the sink. Hence I am wondering if this should be handled outside of the Flink sink and we should just document the expected behavior of the input data to the Flink Iceberg sink.

In CDC case, ff there is some other network shuffle(not base on the primary key) before the FlinkSink, that will lead to cdc record disorder, and yes, we could not get the correct result, because that is out of the control of the FlinkSink. But I think that is user's fault, in CDC case, user should make sure cdc data order correct and send to downstream. 

Besides, we should also to consider the flink default network shuffle and could not manually control network shuffle(such as FlinkSQL, we could not easily to control network shuffle between two opr) before the FlinkSink situation. I think that is Iceberg should resolve, and don't need user to do this manually. User only need to setting correct `equalityFieldColumns` and send cdc/upsert data to FlinkSink, and Iceberg will handle the rest of the work well.

**Comment by stevenzwu** [comment]
@Reo-LEI Thanks a lot for the detailed explanations. I got the problem and motivation for this change. 

> Case2: x != y == z

This PR change the parallelism of of the upstream operator to force x=y. To me, it is dangerous for FlinkSink to modify sth that doesn't own. It violates the principle of ownership/isolation. I would argue that for the CDC upsert case, job parallelism should be set to the CDC source parallelism (x). It doesn't make sense to have a different job parallelism (y) than CDC source parallelism (x). and then have the FlinkSink to do some magic to override the parallelism of an upstream operator that it doesn't own.

> Case3: x == y != z

This can happen if we need an higher parallelism for the Flink writer. I can see that we may want to handle this case in the FlinkSink. I am wondering if we should add a new `equalityKeysHash`  to the `DistributionMode`.

However, I am personally not sure how valuable this setup will be. This assumes that Flink writers are the bottleneck and the job throughput can improve significantly with higher Iceberg writer parallelism.

> Case4: x != y != z

it 's a combination of 2 and 3. so individual arguments apply separately here.

**Comment by Reo-LEI** [comment]
> It doesn't make sense to have a different job parallelism (y) than CDC source parallelism (x). and then have the FlinkSink to do some magic to override the parallelism of an upstream operator that it doesn't own.

But CDC source parallelism (x) and job parallelism(y) are out of iceberg and define by user. Dose iceberg should restrict CDC source parallelism must equal to job parallelism? In other hand, if x != y and we don't override the parallelism of the FlinkSink upstream operator, Is there another way to resolve the disorder problem?

> I am wondering if we should add a new equalityKeysHash to the DistributionMode.

I think we need to discuss the relation between equailty keys and partition keys first. https://github.com/apache/iceberg/pull/2898#discussion_r697882281 @stevenzwu

**Comment by stevenzwu** [comment]
> But CDC source parallelism (x) and job parallelism(y) are out of iceberg and define by user. Dose iceberg should restrict CDC source parallelism must equal to job parallelism?

How does this scenario work for other sinks (like MySQL)? How can other sinks ensure the processing ordering? Do all sinks need to override the parallelism of upstream operator and force the upstream operator parallelism matches the source parallelism? To me, x=y is not a behavior imposed by Iceberg sink. Rather, it is a behavior imposed by the CDC source.

> we need to discuss the relation between equality keys and partition keys. Are partition keys a subset of equality keys?

Agree.  My intuition is that they are independent. But I am also not very clear on that.

cc @openinx @rdblue to get more inputs on the whole discussion

**Comment by Reo-LEI** [comment]
> How does this scenario work for other sinks (like MySQL)? How can other sinks ensure the processing ordering?

The `JdbcDynamicTableSink` and `KafkaDynamicSink` will return `OutputFormatProvider` and `SinkFunctionProvider` as `SinkRuntimeProvider`. These `SinkRuntimeProvider` will accept `OutputFormat` or `SinkFunction` and append sink operator to upstream directly. So there is no `Filter` operator between upstream and sink operator. 

Therefore, the other sinks haven't the Disorder problem. But Iceberg FlinkSink need accept upstream DataStream and append writer, commiter operators to it instead of a single sink operator. So we cann't use the `OutputFormatProvider` and `SinkFunctionProvider` API to avoid the `Filter` operator added between upstream and FlinkSink.

 But Fortunately, I think we could ignore the Disorder problem for FlinkSQL. Because with the upgrade of flink to 1.13(https://github.com/apache/iceberg/issues/2558), The `CommonPhysicalSink` has been remove and refactor to `CommonExecSink`. We can found the `Filter` operator parallelism will be set to be the same as its upstream. So I think we can no longer modify the `Filter` operator parallelism, and resolve the Disorder problem by upgrade flink to 1.13.

PS: https://github.com/apache/flink/blob/5f007ff6c8224c6e5a14de2c79296eb85a22fe1f/flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java#L211
